### PR TITLE
Setup APP

### PR DESCRIPTION
**Pull request by Rushyanth**

### CODE REVIEW

File: app/models.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+	modified_at = db.Column(db.DateTime, default=datetime.now, onupdate=datetime.now)	
+	user_id = db.Column(db.Integer, db.ForeignKey('users.id'))
+
+
+	def __repr__(self):
+		return '<TaskList {}>'.format(self.title)
+	def is_own_task(self, id):
+		if id == current_user.id :
+			return True
+		else:
```
Comment (adityav-verma) [inline app/models.py]: No need of this `else`

File: app/models.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+			return False
+	 
+@login.user_loader
```
Comment (adityav-verma) [inline app/models.py]: Move this function to `app/__init__.py`. This function does not make any sense on `models.py` file.

File: app/routes.py [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+from flask_login import current_user, login_user, login_required
+from app.models import Users, TaskList
+from datetime import datetime
+
+@app.route('/')
+@app.route('/home')
+def home():
+	user = {'username': 'rushi'}
+	return render_template('home.html')
+
```
Comment (adityav-verma) [inline app/routes.py]: What is this route and function???? Remove it.

File: app/routes.py [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+		name = form.name.data 
+		email = form.email.data
+		username = form.username.data
+		password = form.password.data
+
+		user = Users(name=name, email=email, username=username)
+		user.set_password(password)
+		db.session.add(user)
+		db.session.commit()
+
```
Comment (adityav-verma) [inline app/routes.py]: Let's add `try/except` block wherever we do DB operations.

File: app/routes.py [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+		password = form.password.data
+		
+		# Get details from database
+		user = Users.query.filter_by(username=username).first()
+		
+		# Check for username and password
+		if user is None or not user.check_password(password):
+			flash("User doesnt exist or password is incorrect")
+			return redirect(url_for('login_get'))
+		else:
```
Comment (adityav-verma) [inline app/routes.py]: `else` not needed.

File: app/routes.py [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+			login_user(user)								# Authentication of User
+			session['user_id'] = user.id
```
Comment (adityav-verma) [inline app/routes.py]: Why adding this to the `session`? Remove it if not needed.

File: app/routes.py [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+		
+
+# User Home Page
+@app.route('/userhome')
+@login_required
+def userhome():
+	# Get the tasks to homepage
+	# Retreive data
+	tasks = current_user.tasks.all()
+	print(tasks)
```
Comment (adityav-verma) [inline app/routes.py]: Remove `print`, shouldn't be there in production code.

File: app/routes.py [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+		return response
+	else:
+		response = make_response(render_template('userhome.html'))
+		response.headers['Last-Modified'] = datetime.now()
+		response.headers['Cache-Control'] = 'no-store, no-cache, must-revalidate, post-check=0, pre-check=0, max-age=0'
+		response.headers['Pragma'] = 'no-cache'
+		response.headers['Expires'] = '-1'
+		return response
+	# Close cursor
+	cur.close()
```
Comment (adityav-verma) [inline app/routes.py]: What is this `cur.close()`? Remove it if not needed. Also remove the related imports.

File: app/routes.py [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+	
+	if len(tasks) > 0 :
+		# return render_template('userhome.html', tasks = tasks)
+		response = make_response(render_template('userhome.html', tasks = tasks))
+		response.headers['Last-Modified'] = datetime.now()
+		response.headers['Cache-Control'] = 'no-store, no-cache, must-revalidate, post-check=0, pre-check=0, max-age=0'
+		response.headers['Pragma'] = 'no-cache'
+		response.headers['Expires'] = '-1'
+		return response
+	else:
```
Comment (adityav-verma) [inline app/routes.py]: Why is this `else` needed? You are doing the exact same thing as above. Handle it in same code.

**Review by adityav-verma — CHANGES_REQUESTED** [review]

File: app/routes.py [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+		# Get Task Details
+		title = form.title.data
+		body = form.body.data
+		status = False
+
+		# Insert tasks into database
+		task = TaskList(title=title, body=body, status=status, user_id=current_user.id)
+		db.session.add(task)
+		db.session.commit()
+
```
Comment (adityav-verma) [inline app/routes.py]: Let's add `try/except` block around db operations.

File: app/routes.py [hunk 11]
```diff
@@ -0,0 +1,218 @@
...
+@app.route('/addTask', methods = ['POST'])
+@login_required
+def addTask_post():
+	form = AddTaskForm(request.form)
+	if form.validate():
+		form = AddTaskForm(request.form)	
+		# Get Task Details
+		title = form.title.data
+		body = form.body.data
+		status = False
```
Comment (adityav-verma) [inline app/routes.py]: Why is this `status` `False`?

File: app/routes.py [hunk 12]
```diff
@@ -0,0 +1,218 @@
...
+# Edit task
+@app.route('/editTask/<string:id>', methods = ['GET'])
+@login_required
+def editTask_get(id):
+	# Get data
+	task = TaskList.query.filter_by(id=id).first()
+	
+	if task is not None and task.is_own_task(task.user_id):
+		return render_template('editTask.html', task=task)
+	else :
```
Comment (adityav-verma) [inline app/routes.py]: Remove this extra `else`, not needed.

File: app/routes.py [hunk 13]
```diff
@@ -0,0 +1,218 @@
...
+		flash("The task doesnot belong to you", "danger")
+		return redirect(url_for('userhome'))
+
+@app.route('/editTask/<string:id>', methods = ['POST'])
+@login_required
+def editTask_post(id):
+	# Get data
+	task = TaskList.query.filter_by(id=id).first()
+
```
Comment (adityav-verma) [inline app/routes.py]: Add `user_id` owner check to this endpoint too.

File: app/routes.py [hunk 14]
```diff
@@ -0,0 +1,218 @@
...
+	# Get data
+	task = TaskList.query.filter_by(id=id).first()
+	
+	# Get Task Details
+	task.title = request.form['title']
+	task.body = request.form['body']
+	task.status = False
+
+	db.session.add(task)
+	db.session.commit()
```
Comment (adityav-verma) [inline app/routes.py]: `try/except`

File: app/routes.py [hunk 15]
```diff
@@ -0,0 +1,218 @@
...
+
+	flash("Task Updated successfully", "success")
+
+	return redirect(url_for('userhome'))
+
+# Delete Task
+@app.route('/deleteTask/<string:id>', methods=['POST'])
+@login_required
+def deleteTask(id):
+	task = TaskList.query.filter_by(id=id).first()
```
Comment (adityav-verma) [inline app/routes.py]: Add `user_id` owner check here. Any logged in user would be able to delete any task.

File: app/routes.py [hunk 16]
```diff
@@ -0,0 +1,218 @@
...
+	db.session.delete(task)
+	db.session.commit()
```
Comment (adityav-verma) [inline app/routes.py]: `try/except`

File: app/routes.py [hunk 17]
```diff
@@ -0,0 +1,218 @@
...
+
+# Status Of Task
+@app.route('/statusOfTask/<string:status>/<string:id>', methods = ['POST'])
+@login_required
+def statusOfTask(status,id):
+
+	if status == 'False':
+		status = True
+	else:
+		status = False
```
Comment (adityav-verma) [inline app/routes.py]: What are you trying to do here? Please explain.

File: app/routes.py [hunk 18]
```diff
@@ -0,0 +1,218 @@
...
+	
+	task = TaskList.query.filter_by(id=id).first()
```
Comment (adityav-verma) [inline app/routes.py]: `user_id` owner check.

File: app/routes.py [hunk 19]
```diff
@@ -0,0 +1,218 @@
...
+	task.status = status
+	db.session.add(task)
+	db.session.commit()
```
Comment (adityav-verma) [inline app/routes.py]: `try/except`

File: app/routes.py [hunk 20]
```diff
@@ -0,0 +1,218 @@
...
+
+	flash("Status changed succesfully ", "success")
+	return redirect(url_for('userhome'))	
+
+
+# Logout
+@app.route('/logout')
+@login_required
+def logout():
+	session.clear();
```
Comment (adityav-verma) [inline app/routes.py]: Use `Flask-Login` `logout_user()`

File: config.py.sample [hunk 21]
```diff
@@ -0,0 +1,4 @@
+import os
+
+class Config(object):
+    SECRET_KEY = os.environ.get('SECRET_KEY') or 'dummy'
```
Comment (adityav-verma) [inline config.py.sample]: Add all the configs with default values in the sample. I am sure you added more configurations like `SQLAlchemy` etc.

File: app/routes.py [hunk 22]
```diff
@@ -0,0 +1,218 @@
...
+def addTask_get():
+	form = AddTaskForm()
+	return render_template('addTask.html', form=form)
+
+@app.route('/addTask', methods = ['POST'])
+@login_required
+def addTask_post():
+	form = AddTaskForm(request.form)
+	if form.validate():
+		form = AddTaskForm(request.form)
```
Comment (coderadi) [inline app/routes.py]: Where is `form.validate()`?

File: app/routes.py (hunk 13 above)
Comment (adityav-verma) [reply app/routes.py]: And also you should have an EditTask form, and then do `form.validate()`.

**Review by adityav-verma — CHANGES_REQUESTED** [review]

File: app/routes.py (hunk 22 above)
Comment (adityav-verma) [inline app/routes.py]: Remove this `form = AddTaskForm(request.form)`, you have already done this before validate function.

**Review by adityav-verma — CHANGES_REQUESTED** [review]

**Comment by adityav-verma** [comment]
Also add a proper PR description.
## What?
What does the PR do?

## Why?
Why is this PR needed?

File: app/models.py (hunk 1 above)
Comment (Rushyanth) [reply app/models.py]: done

File: app/routes.py (hunk 3 above)
Comment (Rushyanth) [reply app/routes.py]: Its the home page
removed this line
user = {'username': 'rushi'}

File: app/routes.py (hunk 5 above)
Comment (Rushyanth) [reply app/routes.py]: done

File: app/routes.py (hunk 6 above)
Comment (Rushyanth) [reply app/routes.py]: Done

File: app/routes.py (hunk 7 above)
Comment (Rushyanth) [reply app/routes.py]: done

File: app/routes.py (hunk 9 above)
Comment (Rushyanth) [reply app/routes.py]: done

File: app/routes.py (hunk 8 above)
Comment (Rushyanth) [reply app/routes.py]: done

File: app/routes.py (hunk 11 above)
Comment (Rushyanth) [reply app/routes.py]: Initial status of all the newly created tasks is False

File: app/routes.py (hunk 12 above)
Comment (Rushyanth) [reply app/routes.py]: done

File: app/routes.py (hunk 13 above)
Comment (Rushyanth) [reply app/routes.py]: owner check is not needed here.
I didnot use WTF form for Edit Task.

File: app/routes.py (hunk 15 above)
Comment (Rushyanth) [reply app/routes.py]: Not necessary. Another user canot go to that page by typing the link
It automatically redirects.

File: app/routes.py (hunk 17 above)
Comment (Rushyanth) [reply app/routes.py]: Changing the status of the task as done or not Done

File: app/routes.py (hunk 18 above)
Comment (Rushyanth) [reply app/routes.py]: Not necessary. It Automatically redirects on button click.

File: app/routes.py (hunk 20 above)
Comment (Rushyanth) [reply app/routes.py]: Done

File: config.py.sample (hunk 21 above)
Comment (Rushyanth) [reply config.py.sample]: Done

File: app/routes.py (hunk 22 above)
Comment (Rushyanth) [reply app/routes.py]: Done

File: app/routes.py [hunk 23]
```diff
@@ -0,0 +1,235 @@
...
+		try:
+			user = Users(name=name, email=email, username=username)
+			user.set_password(password)
+			db.session.add(user)
+			db.session.commit()
+		except Exception:
+			print("in databaseerror exception block")
+			flash("Could not register", "danger")
+			return redirect(url_for('register_get'))
+		else:
```
Comment (adityav-verma) [inline app/routes.py]: Why this `else`?

**Review by adityav-verma — APPROVED** [review]

**Pull request merged**